File: 01.rb
nums = File.readlines('01.txt', chomp: true).map(&:to_i)

def two_sum(nums, expected)
  a = nums.sort

  low, hi = 0, a.length - 1
  while low < hi
    v1, v2 = a[low], a[hi]
    added = v1 + v2
    if added < expected
      low += 1
    elsif added > expected
      hi -= 1
    else
      return [v1, v2]
    end
  end

  []
end

def three_sum(nums, expected)
  a = nums.sort

  for i in 0..a.length-3 do
    v1 = a[i]
    target = expected - v1

    sub_a = a[(i + 1)..a.length]
    ab = two_sum(sub_a, target)
    if !ab.empty?
      v2,v3 = ab
      return [v1, v2, v3]
    end
  end
  
  []
end

res1 = two_sum(nums, 2020).reduce(:*)
res2 = three_sum(nums, 2020).reduce(:*)
puts '*' * 50
puts res1
puts res2
puts '*' * 50

File: 01.js
let fs = require('fs');

function twoSum(input, expected, startIndex) {
  let i = startIndex
  let j = input.length - 1
  while (i <= j) {
    let added = input[i] + input[j]
    if (added === expected) {
      return [input[i],input[j]]
    } else if (added > expected) {
      j--
    } else {
      i++
    }
  }
  return []
}

function threeSum(input, expected) {
  input.sort((a,b) => a - b)

  for (let idx = 0; idx < input.length - 2; idx++) {
    let toFind = expected - input[idx]

    let res = twoSum(input, toFind, idx + 1)

    let found = res.length !== 0
    if (found) {
      return res.concat([input[idx]])
    }
  }

  return []
}

function main() {
  let input = fs.readFileSync('01.txt', 'utf8')
    .split("\n")
    .map(i => parseInt(i))
    .sort((a,b) => a - b);

  let res1 = twoSum(input, 2020, 0).reduce((a,b) => a*b);
  let res2 = threeSum(input, 2020).reduce((a,b) => a*b);
  console.log('***************');
  console.log(res1);
  console.log(res2);
  console.log('***************');
};
main();